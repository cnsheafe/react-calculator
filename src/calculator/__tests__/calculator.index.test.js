import React from 'react'
import { shallow, configure } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import Calculator from '../calculator.index'

configure({ adapter: new Adapter() })

describe('<Calculator />', () => {
  let wrapper = shallow(<Calculator />)
  let instance = wrapper.instance()

  beforeEach(() => {
    wrapper = shallow(<Calculator />)
    instance = wrapper.instance()
  })

  it('should match the snaphost', () => {
    expect(wrapper).toMatchSnapshot()
  })

  describe('#insertButtonSymbolToBuffer', () => {
    const evt = {
      target: {
        textContent: 7
      },
      preventDefault: jest.fn(),
    }

    it('should append a symbol to the buffer', () => {
      instance.inputRef = {
        current: {
          selectionStart: 1,
          focus: jest.fn(),
        }
      }
      instance.insertButtonSymbolToBuffer(evt)
      expect(wrapper.state().buffer).toBe('7')
      expect(evt.preventDefault).toHaveBeenCalled()
    })
  })

  describe('#appendKeyStrokesToBuffer', () => {
    const evt = {
      target: {
        value: '+'
      }
    }

    it('should update the buffer with the new input value', () => {
      instance.inputRef = {
        current: {
          selectionStart: 10
        }
      }
      instance.insertKeyStrokesToBuffer(evt)
      expect(wrapper.state().buffer).toBe('+')
    })
  })

  describe('#parseAdditionAndSubtraction', () => {
    it('should parse the input and return a new string with addition operations solved', () => {
      const parsed = instance.parseAdditionAndSubtraction('2+2+50')
      expect(parsed).toBe('54')
    })

    it('should parse the input when there is a subtraction operation', () => {
      const parsed = instance.parseAdditionAndSubtraction('-10-5-3')
      expect(parsed).toBe('-18')
    })

    it('should return the input if there is not a valid operation', () => {
      const parsed = instance.parseAdditionAndSubtraction('2+g')
      expect(parsed).toBe('2+g')
    })

    it('should parse the input when there is a mix of addition and subtraction', () => {
      const parsed = instance.parseAdditionAndSubtraction('2-2+72-16')
      expect(parsed).toBe('56')
    })

    it('should parse then input when floating numbers are present', () => {
      const parsed = instance.parseAdditionAndSubtraction('2.2+3.3+4.4')
      expect(parsed).toBe('9.9')
    })
  })

  describe('#parseMultiplicationAndDivision', () => {
    it('should parse the the input when there is a multiplication operation', () => {
      const parsed = instance.parseMultiplicationAndDivision('2*4*7*-1')
      expect(parsed).toBe('-56')
    })

    it('should parse the the input when there is a division operation', () => {
      const parsed = instance.parseMultiplicationAndDivision('100/5/10')
      expect(parsed).toBe('2')
    })

    it('should parse the input when there is a mix of multiplication and division', () => {
      const parsed = instance.parseMultiplicationAndDivision('100/5/10*7')
      expect(parsed).toBe('14')
     
    })
  })

  describe('#parseSquareRoot', () => {
    it('should parse the input when there is an square root', () => {
      const parsed = instance.parseSquareRoot('\u221a100')
      expect(parsed).toBe('10')
    })

    it('should parse input when square root has leading coeffecient', () => {
      const parsed = instance.parseSquareRoot('10\u221a100')
      expect(parsed).toBe('100')
    })
  })

  describe('#parseInput', () => {
    const evt = {
      preventDefault: jest.fn(),
    }
    it('should parse input with mixed operators', () => {
      instance.inputRef = {
        current: {
          focus: jest.fn()
        }
      }

      wrapper.setState({ buffer: '3+3*5-6' })
      instance.parseInput(evt)
      expect(wrapper.state().buffer).toBe('12')
      expect(evt.preventDefault).toHaveBeenCalled()
    })
  })
})