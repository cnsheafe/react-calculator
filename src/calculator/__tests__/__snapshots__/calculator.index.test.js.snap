// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Calculator /> should match the snaphost 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Calculator />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <CalcInput
          buffer=""
          onChange={[Function]}
          total={undefined}
/>,
        <div>
          <CalcButtonArray
                    defaultHandler={[Function]}
                    keyPrefix="calc-0"
                    symbols={
                              Array [
                                        Object {
                                          "text": "C",
                                        },
                                        Object {
                                          "text": "/",
                                        },
                                        Object {
                                          "text": "X",
                                        },
                                        Object {
                                          "text": "DEL",
                                        },
                                      ]
                    }
          />
          <CalcButtonArray
                    defaultHandler={[Function]}
                    keyPrefix="calc-1"
                    symbols={
                              Array [
                                        Object {
                                          "text": "7",
                                        },
                                        Object {
                                          "text": "8",
                                        },
                                        Object {
                                          "text": "9",
                                        },
                                        Object {
                                          "text": "-",
                                        },
                                      ]
                    }
          />
          <CalcButtonArray
                    defaultHandler={[Function]}
                    keyPrefix="calc-2"
                    symbols={
                              Array [
                                        Object {
                                          "text": "4",
                                        },
                                        Object {
                                          "text": "5",
                                        },
                                        Object {
                                          "text": "6",
                                        },
                                        Object {
                                          "text": "+",
                                        },
                                      ]
                    }
          />
          <CalcButtonArray
                    defaultHandler={[Function]}
                    keyPrefix="calc-3"
                    symbols={
                              Array [
                                        Object {
                                          "text": "1",
                                        },
                                        Object {
                                          "text": "2",
                                        },
                                        Object {
                                          "text": "3",
                                        },
                                        Object {
                                          "text": "=",
                                        },
                                      ]
                    }
          />
          <CalcButtonArray
                    defaultHandler={[Function]}
                    keyPrefix="calc-4"
                    symbols={
                              Array [
                                        Object {
                                          "text": "%",
                                        },
                                        Object {
                                          "text": "0",
                                        },
                                        Object {
                                          "text": ".",
                                        },
                                        Object {
                                          "text": "√",
                                        },
                                      ]
                    }
          />
</div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "buffer": "",
          "onChange": [Function],
          "total": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <CalcButtonArray
              defaultHandler={[Function]}
              keyPrefix="calc-0"
              symbols={
                            Array [
                                          Object {
                                            "text": "C",
                                          },
                                          Object {
                                            "text": "/",
                                          },
                                          Object {
                                            "text": "X",
                                          },
                                          Object {
                                            "text": "DEL",
                                          },
                                        ]
              }
/>,
            <CalcButtonArray
              defaultHandler={[Function]}
              keyPrefix="calc-1"
              symbols={
                            Array [
                                          Object {
                                            "text": "7",
                                          },
                                          Object {
                                            "text": "8",
                                          },
                                          Object {
                                            "text": "9",
                                          },
                                          Object {
                                            "text": "-",
                                          },
                                        ]
              }
/>,
            <CalcButtonArray
              defaultHandler={[Function]}
              keyPrefix="calc-2"
              symbols={
                            Array [
                                          Object {
                                            "text": "4",
                                          },
                                          Object {
                                            "text": "5",
                                          },
                                          Object {
                                            "text": "6",
                                          },
                                          Object {
                                            "text": "+",
                                          },
                                        ]
              }
/>,
            <CalcButtonArray
              defaultHandler={[Function]}
              keyPrefix="calc-3"
              symbols={
                            Array [
                                          Object {
                                            "text": "1",
                                          },
                                          Object {
                                            "text": "2",
                                          },
                                          Object {
                                            "text": "3",
                                          },
                                          Object {
                                            "text": "=",
                                          },
                                        ]
              }
/>,
            <CalcButtonArray
              defaultHandler={[Function]}
              keyPrefix="calc-4"
              symbols={
                            Array [
                                          Object {
                                            "text": "%",
                                          },
                                          Object {
                                            "text": "0",
                                          },
                                          Object {
                                            "text": ".",
                                          },
                                          Object {
                                            "text": "√",
                                          },
                                        ]
              }
/>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "defaultHandler": [Function],
              "keyPrefix": "calc-0",
              "symbols": Array [
                Object {
                  "text": "C",
                },
                Object {
                  "text": "/",
                },
                Object {
                  "text": "X",
                },
                Object {
                  "text": "DEL",
                },
              ],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "defaultHandler": [Function],
              "keyPrefix": "calc-1",
              "symbols": Array [
                Object {
                  "text": "7",
                },
                Object {
                  "text": "8",
                },
                Object {
                  "text": "9",
                },
                Object {
                  "text": "-",
                },
              ],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "defaultHandler": [Function],
              "keyPrefix": "calc-2",
              "symbols": Array [
                Object {
                  "text": "4",
                },
                Object {
                  "text": "5",
                },
                Object {
                  "text": "6",
                },
                Object {
                  "text": "+",
                },
              ],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "defaultHandler": [Function],
              "keyPrefix": "calc-3",
              "symbols": Array [
                Object {
                  "text": "1",
                },
                Object {
                  "text": "2",
                },
                Object {
                  "text": "3",
                },
                Object {
                  "text": "=",
                },
              ],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "defaultHandler": [Function],
              "keyPrefix": "calc-4",
              "symbols": Array [
                Object {
                  "text": "%",
                },
                Object {
                  "text": "0",
                },
                Object {
                  "text": ".",
                },
                Object {
                  "text": "√",
                },
              ],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <CalcInput
            buffer=""
            onChange={[Function]}
            total={undefined}
/>,
          <div>
            <CalcButtonArray
                        defaultHandler={[Function]}
                        keyPrefix="calc-0"
                        symbols={
                                    Array [
                                                Object {
                                                  "text": "C",
                                                },
                                                Object {
                                                  "text": "/",
                                                },
                                                Object {
                                                  "text": "X",
                                                },
                                                Object {
                                                  "text": "DEL",
                                                },
                                              ]
                        }
            />
            <CalcButtonArray
                        defaultHandler={[Function]}
                        keyPrefix="calc-1"
                        symbols={
                                    Array [
                                                Object {
                                                  "text": "7",
                                                },
                                                Object {
                                                  "text": "8",
                                                },
                                                Object {
                                                  "text": "9",
                                                },
                                                Object {
                                                  "text": "-",
                                                },
                                              ]
                        }
            />
            <CalcButtonArray
                        defaultHandler={[Function]}
                        keyPrefix="calc-2"
                        symbols={
                                    Array [
                                                Object {
                                                  "text": "4",
                                                },
                                                Object {
                                                  "text": "5",
                                                },
                                                Object {
                                                  "text": "6",
                                                },
                                                Object {
                                                  "text": "+",
                                                },
                                              ]
                        }
            />
            <CalcButtonArray
                        defaultHandler={[Function]}
                        keyPrefix="calc-3"
                        symbols={
                                    Array [
                                                Object {
                                                  "text": "1",
                                                },
                                                Object {
                                                  "text": "2",
                                                },
                                                Object {
                                                  "text": "3",
                                                },
                                                Object {
                                                  "text": "=",
                                                },
                                              ]
                        }
            />
            <CalcButtonArray
                        defaultHandler={[Function]}
                        keyPrefix="calc-4"
                        symbols={
                                    Array [
                                                Object {
                                                  "text": "%",
                                                },
                                                Object {
                                                  "text": "0",
                                                },
                                                Object {
                                                  "text": ".",
                                                },
                                                Object {
                                                  "text": "√",
                                                },
                                              ]
                        }
            />
</div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "buffer": "",
            "onChange": [Function],
            "total": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <CalcButtonArray
                defaultHandler={[Function]}
                keyPrefix="calc-0"
                symbols={
                                Array [
                                                Object {
                                                  "text": "C",
                                                },
                                                Object {
                                                  "text": "/",
                                                },
                                                Object {
                                                  "text": "X",
                                                },
                                                Object {
                                                  "text": "DEL",
                                                },
                                              ]
                }
/>,
              <CalcButtonArray
                defaultHandler={[Function]}
                keyPrefix="calc-1"
                symbols={
                                Array [
                                                Object {
                                                  "text": "7",
                                                },
                                                Object {
                                                  "text": "8",
                                                },
                                                Object {
                                                  "text": "9",
                                                },
                                                Object {
                                                  "text": "-",
                                                },
                                              ]
                }
/>,
              <CalcButtonArray
                defaultHandler={[Function]}
                keyPrefix="calc-2"
                symbols={
                                Array [
                                                Object {
                                                  "text": "4",
                                                },
                                                Object {
                                                  "text": "5",
                                                },
                                                Object {
                                                  "text": "6",
                                                },
                                                Object {
                                                  "text": "+",
                                                },
                                              ]
                }
/>,
              <CalcButtonArray
                defaultHandler={[Function]}
                keyPrefix="calc-3"
                symbols={
                                Array [
                                                Object {
                                                  "text": "1",
                                                },
                                                Object {
                                                  "text": "2",
                                                },
                                                Object {
                                                  "text": "3",
                                                },
                                                Object {
                                                  "text": "=",
                                                },
                                              ]
                }
/>,
              <CalcButtonArray
                defaultHandler={[Function]}
                keyPrefix="calc-4"
                symbols={
                                Array [
                                                Object {
                                                  "text": "%",
                                                },
                                                Object {
                                                  "text": "0",
                                                },
                                                Object {
                                                  "text": ".",
                                                },
                                                Object {
                                                  "text": "√",
                                                },
                                              ]
                }
/>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "defaultHandler": [Function],
                "keyPrefix": "calc-0",
                "symbols": Array [
                  Object {
                    "text": "C",
                  },
                  Object {
                    "text": "/",
                  },
                  Object {
                    "text": "X",
                  },
                  Object {
                    "text": "DEL",
                  },
                ],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "defaultHandler": [Function],
                "keyPrefix": "calc-1",
                "symbols": Array [
                  Object {
                    "text": "7",
                  },
                  Object {
                    "text": "8",
                  },
                  Object {
                    "text": "9",
                  },
                  Object {
                    "text": "-",
                  },
                ],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "defaultHandler": [Function],
                "keyPrefix": "calc-2",
                "symbols": Array [
                  Object {
                    "text": "4",
                  },
                  Object {
                    "text": "5",
                  },
                  Object {
                    "text": "6",
                  },
                  Object {
                    "text": "+",
                  },
                ],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "defaultHandler": [Function],
                "keyPrefix": "calc-3",
                "symbols": Array [
                  Object {
                    "text": "1",
                  },
                  Object {
                    "text": "2",
                  },
                  Object {
                    "text": "3",
                  },
                  Object {
                    "text": "=",
                  },
                ],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "defaultHandler": [Function],
                "keyPrefix": "calc-4",
                "symbols": Array [
                  Object {
                    "text": "%",
                  },
                  Object {
                    "text": "0",
                  },
                  Object {
                    "text": ".",
                  },
                  Object {
                    "text": "√",
                  },
                ],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
